### PR TITLE
SDK Team Guidance around Releases

### PR DESCRIPTION
**Pull request by QuintinWillison**
While we're working on moving more of our SDK repository release processes to execute 'from the cloud', it's become clear that we need a clear template for how this should ideally be done.

The process outlined in this template needs to accommodate the way that we like to work in the SDK Team at Ably. This is just a proposal and, therefore, it is open and expected to change as the views of team members come in. We also expect this first iteration of this template to evolve over time as we learn more about how we like to do releases or technologies change which require us to change our stance.

Creation of this document was triggered by [this comment](https://github.com/ably/ably-js/issues/805#issuecomment-1067856291), by @jamienewcomb, asking for acceptance criteria for `ably-js` when implementing 'Continuous Delivery via GitHub workflow'.

While this pull request remains open, the rendered output can be previewed [on the underlying branch](https://github.com/ably/engineering/blob/sdk-releases/sdk/releases.md).

### CODE REVIEW

File: sdk/releases.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+- where it's necessary to deviate from some of the ideals specified here due to constraints which are innate or otherwise idiomatic to the platform being released for
+
+Where technically possible, SDK repository release processes which claim to adhere to this guidance must follow these ideal requirements closely.
+
+## Release Process
+
+1. Merge all pull requests containing changes intended for this release to `main` branch
+2. Prepare a [Release Branch](#release-branch) and a corresponding pull request, obtain approval from reviewers and then merge to `main` branch
+3. Trigger the [Publish Workflow](#publish-workflow)
+4. If the publish fails then add commits to the `main` branch that aim to fix the problem and then try step 3 again
```
Comment (lmars) [inline sdk/releases.md:17]: I'm not sure how I feel about direct pushes to `main`, I think we should always encourage changes to be peer reviewed, or at least a PR + merge with an explicit comment justifying why the change was merged without a review.

File: sdk/releases.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+5. Push Git version tag, which may or may not have a `v` prefix according to the existing conventions for this SDK repository
```
Comment (lmars) [inline sdk/releases.md:18]: Is it worth just standardising this one way or the other? Or or there some specific reasons why some SDKs require the prefix, and others don't?

File: sdk/releases.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+6. Create GitHub release
+
+## Release Branch
+
+- Branch from the `main` branch
+- Merge to the `main` branch, once approved
+- Named liked `release/<version>`
+- Increment the version
+- Version should conform to [SemVer](https://semver.org/)
+- Add change log entry
```
Comment (lmars) [inline sdk/releases.md:28]: Should the process for this be documented and linked to?

File: sdk/releases.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+## Publish Workflow
+
+- Run in GitHub environment
+- Be manually triggered using a [`workflow_dispatch` event](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_dispatch)
+- Override the default [checkout](https://github.com/actions/checkout) `ref` with a commit SHA supplied as in input to the triggering event, where that SHA will generally be `HEAD` of the `main` branch, at the point the [Release Branch](#release-branch) was merged
+- Publish to downstream package repositories using an Ably identity, not using an identity tied to an individual person in any way
+- Use GitHub repository secrets to securely store the credentials required to publish under the Ably identity
```
Comment (lmars) [inline sdk/releases.md:36]: We should also consider using [GitHub OIDC](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect) where appropriate rather than managing secrets, we recently put this in place for some internal GitHub actions to retrieve short lived AWS credentials. I'd be happy to discuss this further.

File: sdk/releases.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+3. Trigger the [Publish Workflow](#publish-workflow)
+4. If the publish fails then add commits to the `main` branch that aim to fix the problem and then try step 3 again
+5. Push Git version tag, which may or may not have a `v` prefix according to the existing conventions for this SDK repository
+6. Create GitHub release
+
+## Release Branch
+
+- Branch from the `main` branch
+- Merge to the `main` branch, once approved
+- Named liked `release/<version>`
```
Comment (lawrence-forooghian) [inline sdk/releases.md:25]:
```suggestion
- Named like `release/<version>`
```

File: sdk/releases.md (hunk 2 above)
Comment (lawrence-forooghian) [inline sdk/releases.md:18]: Does the tag & release not need to come before the publish workflow? I'm not familiar with the details of how publishing to a package repository works, but I imagine that in some cases we need to refer to an existing tag when we publish?

File: sdk/releases.md (hunk 1 above)
Comment (paddybyers) [reply sdk/releases.md:17]: I agree - branch protection should be in place so this isn't possible

Comment (QuintinWillison) [reply sdk/releases.md:17]: Thanks, both. I worded that really badly. I would prefer that these went via pull requests and should have made that clearer. I'll work on improving this phrasing.

File: sdk/releases.md (hunk 2 above)
Comment (QuintinWillison) [reply sdk/releases.md:18]: Honestly, I don't know. This is all stuff that was put in place before I started working for Ably, so there's history needs digging into. I would _love_ to conform everywhere if possible - it's just sometimes impossible to tell what downstream dependencies might be reliant on current naming conventions per repository.

File: sdk/releases.md (hunk 3 above)
Comment (QuintinWillison) [reply sdk/releases.md:28]: Yes. I will expand on this.

File: sdk/releases.md (hunk 4 above)
Comment (QuintinWillison) [reply sdk/releases.md:36]: I'll add mention, thanks. Falls outside scope of this PR, otherwise.

File: sdk/releases.md (hunk 2 above)
Comment (QuintinWillison) [reply sdk/releases.md:18]: I recognise that there will be instances, hence [this statement](https://github.com/ably/engineering/blob/e712b93b53342ad51b493391c26e4e5ec1ec9719/sdk/releases.md?plain=1#L8) earlier on in this document:

> where it's necessary to deviate from some of the ideals specified here due to constraints which are innate or otherwise idiomatic to the platform being released for

However, generally, our preference should be to successfully make the release _before_ pushing the tag. This avoids any requirement for us to do messy things like moving tags.

File: sdk/releases.md (hunk 1 above)
Comment (QuintinWillison) [reply sdk/releases.md]: https://github.com/ably/engineering/pull/23/commits/5a0f0d8236d2b215cea105e3b3a6c3b6d5679634

File: sdk/releases.md (hunk 3 above)
Comment (QuintinWillison) [reply sdk/releases.md]: https://github.com/ably/engineering/pull/23/commits/008beed2c2d97a8876122a531aa1f8f314cdeeca

File: sdk/releases.md (hunk 4 above)
Comment (QuintinWillison) [reply sdk/releases.md]: https://github.com/ably/engineering/pull/23/commits/649a4f37958eaf80d10c2273a13d671d73e82321

File: sdk/releases.md [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+5. Push Git version tag, which may or may not have a `v` prefix according to the existing conventions for this SDK repository
+6. Create GitHub release
+
+## Release Branch
+
+- Branch from the `main` branch
+- Merge to the `main` branch, once approved
+- Named like `release/<version>`
+- Increment the version
+- Version should conform to [SemVer](https://semver.org/)
```
[the remark this answers is not in the document]
Comment (QuintinWillison) [reply sdk/releases.md:27]: Good point. This pull request represents where we _want to be_. So, yes indeed, we currently have a state of play across all SDKs where our version numbers look like they conform to SemVer but are constrained to utilising only `patch` increments. The expectation is that this constraint will be unblocked very soon (cc @stmoreau), therefore I think it's appropriate that the scope of this pull request clarifies our _ideal_ configuration (as outlined at the top of this document).

In due course, once we are using SemVer properly across our SDKs (i.e. incrementing `major`, `minor` and `patch` appropriately to communicate correctly to human and automated package resolvers alike) then I would like to expand on this. That falls outside the scope of this pull request.

File: sdk/releases.md (hunk 2 above)
Comment (lawrence-forooghian) [reply sdk/releases.md:18]: OK, cool 👍

Comment (lmars) [reply sdk/releases.md:18]: Ok, so is it worth at least having an opinion here which one is more preferred so we're at least converging on one option going forwards?

File: sdk/releases.md [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+- Add change log entry (process to be documented under [#17](https://github.com/ably/engineering/issues/17))
+
+## Publish Workflow
+
+- Run in GitHub environment
+- Be manually triggered using a [`workflow_dispatch` event](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_dispatch)
```
Comment (ikbalkaya) [inline sdk/releases.md:33]: I found 'Be manually triggered'. a bit confusing. Would it be a bit clearer if worded something like 'configured to be manually triggered' 🤔

File: sdk/releases.md (hunk 2 above)
Comment (QuintinWillison) [reply sdk/releases.md]: https://github.com/ably/engineering/pull/23/commits/edc9d8d54d48ffeb42fcf959ff253cae12fd41af

Consensus was clear from the poll that we should prefer the `v` prefix. As @owenpearson pointed out [in that internal Slack thread](https://ably-real-time.slack.com/archives/C030C5YLY/p1647963277793149?thread_ts=1647963003.830159&cid=C030C5YLY):

> it makes sense to take conventions from the creator of git himself https://github.com/torvalds/linux/tags

**Review by stmoreau — APPROVED** [review]

**Review by AndyNicks — APPROVED** [review]

File: sdk/releases.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+- Branch from the `main` branch
+- Merge to the `main` branch, once approved
+- Named like `release/<version>`
+- Increment the version
+- Version should conform to [SemVer](https://semver.org/)
+- Add change log entry (process to be documented under [#17](https://github.com/ably/engineering/issues/17))
+
+## Publish Workflow
+
+- Run in GitHub environment
```
Comment (owenpearson) [inline sdk/releases.md:32]: There's an implicit "should" or "must" at the beginning of each of these sentences and we should make that explicit. I see you've prefaced the list under 'Version Tag' with 'Should:' so maybe we should do the same here?

**Review by owenpearson — APPROVED** [review]

**Review by Peter-Maguire — APPROVED** [review]

File: sdk/releases.md (hunk 7 above)
Comment (QuintinWillison) [reply sdk/releases.md]: https://github.com/ably/engineering/pull/23/commits/baa07b07ad332489f80bbfa02e81bde43d9e38dd

**Review by ikbalkaya — APPROVED** [review]

**Review by tomkirbygreen — APPROVED** [review]

**Review by Rosalita — APPROVED** [review]

**Review by lmars — APPROVED** [review]

**Review by lawrence-forooghian — APPROVED** [review]

**Review by paddybyers — APPROVED** [review]
LGTM, thanks
